File: manuscript/Dataset_8_PDGF_signaling_slows_ovarian_cancer GSM7019822_GSM7019823_GSM7019824/new_analysis/test1.py
import numpy as np
from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister
from qiskit.circuit.library import RYGate
from qiskit_aer import AerSimulator
from qiskit import transpile
import matplotlib.pyplot as plt

def create_state_preparation_circuit(frequencies):
    """
    Create a quantum circuit that prepares a state with given measurement frequencies.
    Uses a systematic approach with controlled rotations.
    """
    # Normalize frequencies to probabilities
    frequencies = np.array(frequencies)
    probabilities = frequencies / np.sum(frequencies)
    
    # Convert to amplitudes (assuming real amplitudes)
    amplitudes = np.sqrt(probabilities)
    
    print("Input frequencies:", frequencies)
    print("Normalized probabilities:", probabilities)
    print("Target amplitudes:", amplitudes)
    
    # Create quantum circuit with 3 qubits
    qc = QuantumCircuit(3, 3)
    
    # Method: Uniform Controlled Rotations
    # We'll use a systematic approach to prepare the state
    
    # First, let's group the amplitudes by qubit patterns
    # State |abc⟩ corresponds to amplitudes[4*a + 2*b + c]
    
    # Calculate rotation angles for each level
    # Level 1: Split probability between first 4 states (000-011) and last 4 states (100-111)
    prob_first_half = np.sum(probabilities[:4])
    prob_second_half = np.sum(probabilities[4:])
    
    if prob_first_half + prob_second_half > 0:
        theta_0 = 2 * np.arcsin(np.sqrt(prob_second_half))
        qc.ry(theta_0, 0)
    
    # Level 2: For each half, split between first 2 and last 2 states
    # First half: states 000-001 vs 010-011
    if prob_first_half > 0:
        prob_00x = np.sum(probabilities[0:2])
        prob_01x = np.sum(probabilities[2:4])
        if prob_00x + prob_01x > 0:
            theta_1 = 2 * np.arcsin(np.sqrt(prob_01x / (prob_00x + prob_01x)))
            qc.cry(theta_1, 0, 1)
    
    # Second half: states 100-101 vs 110-111
    if prob_second_half > 0:
        prob_10x = np.sum(probabilities[4:6])
        prob_11x = np.sum(probabilities[6:8])
        if prob_10x + prob_11x > 0:
            theta_2 = 2 * np.arcsin(np.sqrt(prob_11x / (prob_10x + prob_11x)))
            qc.cry(theta_2, 0, 1)
    
    # Level 3: For each quarter, split between individual states
    # We'll implement controlled-controlled-RY using basic gates
    
    # Quarter 1: 000 vs 001 (control on q0=0, q1=0)
    if np.sum(probabilities[0:2]) > 0:
        theta_3 = 2 * np.arcsin(np.sqrt(probabilities[1] / np.sum(probabilities[0:2])))
        qc.x(0)  # flip q0 to make it |1⟩
        qc.x(1)  # flip q1 to make it |1⟩
        qc.ccx(0, 1, 2)  # Toffoli gate as control
        qc.ry(theta_3, 2)  # Apply rotation
        qc.ccx(0, 1, 2)  # Undo Toffoli
        qc.x(1)  # restore q1
        qc.x(0)  # restore q0
    
    # Quarter 2: 010 vs 011 (control on q0=0, q1=1)
    if np.sum(probabilities[2:4]) > 0:
        theta_4 = 2 * np.arcsin(np.sqrt(probabilities[3] / np.sum(probabilities[2:4])))
        qc.x(0)  # flip q0 to make it |1⟩
        qc.ccx(0, 1, 2)  # Toffoli gate as control
        qc.ry(theta_4, 2)  # Apply rotation
        qc.ccx(0, 1, 2)  # Undo Toffoli
        qc.x(0)  # restore q0
    
    # Quarter 3: 100 vs 101 (control on q0=1, q1=0)
    if np.sum(probabilities[4:6]) > 0:
        theta_5 = 2 * np.arcsin(np.sqrt(probabilities[5] / np.sum(probabilities[4:6])))
        qc.x(1)  # flip q1 to make it |1⟩
        qc.ccx(0, 1, 2)  # Toffoli gate as control
        qc.ry(theta_5, 2)  # Apply rotation
        qc.ccx(0, 1, 2)  # Undo Toffoli
        qc.x(1)  # restore q1
    
    # Quarter 4: 110 vs 111 (control on q0=1, q1=1)
    if np.sum(probabilities[6:8]) > 0:
        theta_6 = 2 * np.arcsin(np.sqrt(probabilities[7] / np.sum(probabilities[6:8])))
        qc.ccx(0, 1, 2)  # Toffoli gate as control
        qc.ry(theta_6, 2)  # Apply rotation
        qc.ccx(0, 1, 2)  # Undo Toffoli
    
    return qc, amplitudes

def simulate_circuit(qc, shots=10000):
    """Simulate the quantum circuit and return measurement frequencies."""
    # Add measurements
    qc_with_measurement = qc.copy()
    qc_with_measurement.measure_all()
    
    # Simulate
    simulator = AerSimulator()
    compiled_circuit = transpile(qc_with_measurement, simulator)
    result = simulator.run(compiled_circuit, shots=shots).result()
    counts = result.get_counts()
    
    # Convert to frequency array
    frequencies = np.zeros(8)
    for state, count in counts.items():
        # Remove spaces and convert binary string to decimal index
        clean_state = state.replace(' ', '')
        # Reverse the string because Qiskit uses little-endian bit ordering
        # (rightmost bit is qubit 0)
        reversed_state = clean_state[::-1]
        index = int(reversed_state, 2)
        frequencies[index] = count / shots
    
    return frequencies

def alternative_method(frequencies):
    """
    Alternative method using a more direct approach with RY gates.
    This method systematically builds the superposition.
    """
    probabilities = np.array(frequencies) / np.sum(frequencies)
    
    qc = QuantumCircuit(3, 3)
    
    # More systematic approach using the uniform controlled rotations
    # This is a simplified version - for exact matching, we'd need optimization
    
    # Calculate cumulative probabilities for systematic preparation
    cumprobs = np.cumsum(probabilities)
    
    # Apply rotations to approximate the target distribution
    # This is a heuristic approach
    
    # First qubit rotation
    p_first_half = np.sum(probabilities[:4])
    if p_first_half < 1.0:
        angle = 2 * np.arccos(np.sqrt(p_first_half))
        qc.ry(angle, 0)
    
    # Second qubit rotations (controlled)
    p_00 = np.sum(probabilities[0:2])
    p_01 = np.sum(probabilities[2:4])
    if p_00 + p_01 > 0:
        angle = 2 * np.arccos(np.sqrt(p_00 / (p_00 + p_01)))
        qc.cry(angle, 0, 1)
    
    p_10 = np.sum(probabilities[4:6])
    p_11 = np.sum(probabilities[6:8])
    if p_10 + p_11 > 0:
        angle = 2 * np.arccos(np.sqrt(p_10 / (p_10 + p_11)))
        qc.x(0)
        qc.cry(angle, 0, 1)
        qc.x(0)
    
    # Third qubit rotations (doubly controlled using basic gates)
    # Helper function to implement controlled-controlled-RY
    def ccry(qc, angle, control1, control2, target):
        """Implement controlled-controlled-RY using basic gates"""
        qc.cry(angle/2, control2, target)
        qc.cx(control1, control2)
        qc.cry(-angle/2, control2, target)
        qc.cx(control1, control2)
        qc.cry(angle/2, control1, target)
    
    if probabilities[0] + probabilities[1] > 0:
        angle = 2 * np.arccos(np.sqrt(probabilities[0] / (probabilities[0] + probabilities[1])))
        qc.x(0)
        qc.x(1)
        ccry(qc, angle, 0, 1, 2)
        qc.x(1)
        qc.x(0)
    
    if probabilities[2] + probabilities[3] > 0:
        angle = 2 * np.arccos(np.sqrt(probabilities[2] / (probabilities[2] + probabilities[3])))
        qc.x(0)
        ccry(qc, angle, 0, 1, 2)
        qc.x(0)
    
    if probabilities[4] + probabilities[5] > 0:
        angle = 2 * np.arccos(np.sqrt(probabilities[4] / (probabilities[4] + probabilities[5])))
        qc.x(1)
        ccry(qc, angle, 0, 1, 2)
        qc.x(1)
    
    if probabilities[6] + probabilities[7] > 0:
        angle = 2 * np.arccos(np.sqrt(probabilities[6] / (probabilities[6] + probabilities[7])))
        ccry(qc, angle, 0, 1, 2)
    
    return qc

# Your input frequencies
target_frequencies = [0.2346, 0.1829, 0.0911, 0.0617, 0.1646, 0.1259, 0.0807, 0.0585]

# Create the quantum circuit
print("Creating quantum state preparation circuit...")
qc, target_amplitudes = create_state_preparation_circuit(target_frequencies)

print(f"\nQuantum Circuit:")
print(qc)

# Simulate the circuit
print("\nSimulating circuit...")
measured_frequencies = simulate_circuit(qc, shots=10000)

# Compare results
print("\nComparison of frequencies:")
print("State | Target   | Measured | Error")
print("------|----------|----------|--------")
for i in range(8):
    binary = format(i, '03b')
    error = abs(target_frequencies[i] - measured_frequencies[i])
    print(f"|{binary}⟩  | {target_frequencies[i]:.4f}   | {measured_frequencies[i]:.4f}   | {error:.4f}")

total_error = np.sum(np.abs(np.array(target_frequencies) - measured_frequencies))
print(f"\nTotal absolute error: {total_error:.4f}")

# Try alternative method
print("\n" + "="*50)
print("Trying alternative method...")
qc_alt = alternative_method(target_frequencies)
print(f"\nAlternative Circuit:")
print(qc_alt)

measured_frequencies_alt = simulate_circuit(qc_alt, shots=10000)
print("\nAlternative method results:")
print("State | Target   | Measured | Error")
print("------|----------|----------|--------")
for i in range(8):
    binary = format(i, '03b')
    error = abs(target_frequencies[i] - measured_frequencies_alt[i])
    print(f"|{binary}⟩  | {target_frequencies[i]:.4f}   | {measured_frequencies_alt[i]:.4f}   | {error:.4f}")

total_error_alt = np.sum(np.abs(np.array(target_frequencies) - measured_frequencies_alt))
print(f"\nTotal absolute error (alternative): {total_error_alt:.4f}")

File: manuscript/Dataset_8_PDGF_signaling_slows_ovarian_cancer GSM7019822_GSM7019823_GSM7019824/new_analysis/test1.m
function quantum_state_preparation()
    % Quantum State Preparation in MATLAB
    % Creates a 3-qubit quantum circuit to match given frequency distribution
    
    % Your target frequencies
    target_frequencies = [0.2346, 0.1829, 0.0911, 0.0617, 0.1646, 0.1259, 0.0807, 0.0585];
    
    fprintf('Quantum State Preparation for 3 Qubits\n');
    fprintf('=====================================\n');
    
    % Method 1: Direct amplitude encoding
    fprintf('\nMethod 1: Direct Amplitude Encoding\n');
    [circuit1, state1] = create_amplitude_encoding_circuit(target_frequencies);
    measured_freq1 = simulate_measurements(state1, 10000);
    
    % Method 2: Systematic controlled rotations
    fprintf('\nMethod 2: Systematic Controlled Rotations\n');
    [circuit2, state2] = create_systematic_circuit(target_frequencies);
    measured_freq2 = simulate_measurements(state2, 10000);
    
    % Compare results
    fprintf('\nComparison of Results:\n');
    fprintf('State | Target   | Method1  | Method2  | Error1   | Error2\n');
    fprintf('------|----------|----------|----------|----------|----------\n');
    
    total_error1 = 0;
    total_error2 = 0;
    
    for i = 1:8
        binary = dec2bin(i-1, 3);
        error1 = abs(target_frequencies(i) - measured_freq1(i));
        error2 = abs(target_frequencies(i) - measured_freq2(i));
        total_error1 = total_error1 + error1;
        total_error2 = total_error2 + error2;
        
        fprintf('|%s⟩  | %.4f   | %.4f   | %.4f   | %.4f   | %.4f\n', ...
            binary, target_frequencies(i), measured_freq1(i), measured_freq2(i), error1, error2);
    end
    
    fprintf('\nTotal Absolute Errors:\n');
    fprintf('Method 1: %.4f\n', total_error1);
    fprintf('Method 2: %.4f\n', total_error2);
    
    % Visualize results
    visualize_results(target_frequencies, measured_freq1, measured_freq2);
end

function [circuit, final_state] = create_amplitude_encoding_circuit(frequencies)
    % Direct amplitude encoding method
    % Converts frequencies directly to quantum amplitudes
    
    % Normalize to probabilities
    probabilities = frequencies / sum(frequencies);
    
    % Convert to amplitudes (assuming real amplitudes)
    amplitudes = sqrt(probabilities);
    
    % Create initial state |000⟩
    initial_state = [1; 0; 0; 0; 0; 0; 0; 0];
    
    % The target state is directly the amplitude vector
    final_state = amplitudes';
    
    % Circuit description (conceptual)
    circuit = struct();
    circuit.method = 'Direct Amplitude Encoding';
    circuit.description = 'Directly sets amplitudes to sqrt(probabilities)';
    circuit.gates = {'State Preparation'};
    
    fprintf('Target amplitudes: [%.4f, %.4f, %.4f, %.4f, %.4f, %.4f, %.4f, %.4f]\n', ...
        amplitudes);
end

function [circuit, final_state] = create_systematic_circuit(frequencies)
    % Systematic approach using controlled rotations
    % Builds the state hierarchically
    
    % Normalize to probabilities
    probabilities = frequencies / sum(frequencies);
    
    % Start with |000⟩ state
    state = [1; 0; 0; 0; 0; 0; 0; 0];
    
    % Circuit structure
    circuit = struct();
    circuit.method = 'Systematic Controlled Rotations';
    circuit.gates = {};
    gate_count = 0;
    
    % Level 1: Split between first 4 states (000-011) and last 4 states (100-111)
    prob_first_half = sum(probabilities(1:4));
    prob_second_half = sum(probabilities(5:8));
    
    if prob_first_half + prob_second_half > 0
        theta_0 = 2 * asin(sqrt(prob_second_half));
        state = apply_ry_gate(state, 1, theta_0);  % qubit 0 (1st qubit)
        gate_count = gate_count + 1;
        circuit.gates{gate_count} = sprintf('RY(%.4f) on qubit 0', theta_0);
    end
    
    % Level 2: Split within each half
    % First half: 000-001 vs 010-011
    if prob_first_half > 0
        prob_00x = sum(probabilities(1:2));
        prob_01x = sum(probabilities(3:4));
        if prob_00x + prob_01x > 0
            theta_1 = 2 * asin(sqrt(prob_01x / (prob_00x + prob_01x)));
            state = apply_controlled_ry_gate(state, 1, 2, theta_1);  % control: qubit 0, target: qubit 1
            gate_count = gate_count + 1;
            circuit.gates{gate_count} = sprintf('CRY(%.4f) control=0, target=1', theta_1);
        end
    end
    
    % Second half: 100-101 vs 110-111
    if prob_second_half > 0
        prob_10x = sum(probabilities(5:6));
        prob_11x = sum(probabilities(7:8));
        if prob_10x + prob_11x > 0
            theta_2 = 2 * asin(sqrt(prob_11x / (prob_10x + prob_11x)));
            state = apply_controlled_ry_gate(state, 1, 2, theta_2);  % control: qubit 0, target: qubit 1
            gate_count = gate_count + 1;
            circuit.gates{gate_count} = sprintf('CRY(%.4f) control=0, target=1', theta_2);
        end
    end
    
    % Level 3: Split individual states within each quarter
    quarters = {[1,2], [3,4], [5,6], [7,8]};
    control_patterns = {[0,0], [0,1], [1,0], [1,1]};
    
    for q = 1:4
        indices = quarters{q};
        prob_quarter = sum(probabilities(indices));
        
        if prob_quarter > 0 && probabilities(indices(1)) + probabilities(indices(2)) > 0
            theta = 2 * asin(sqrt(probabilities(indices(2)) / (probabilities(indices(1)) + probabilities(indices(2)))));
            
            % Apply doubly controlled rotation
            state = apply_ccry_gate(state, control_patterns{q}, theta);
            gate_count = gate_count + 1;
            circuit.gates{gate_count} = sprintf('CCRY(%.4f) controls=[%d,%d], target=2', ...
                theta, control_patterns{q}(1), control_patterns{q}(2));
        end
    end
    
    final_state = state;
    
    fprintf('Circuit has %d gates\n', gate_count);
end

function new_state = apply_ry_gate(state, qubit, angle)
    % Apply RY rotation to specified qubit
    % qubit: 1, 2, or 3 (1-indexed)
    
    % RY gate matrix
    RY = [cos(angle/2), -sin(angle/2); sin(angle/2), cos(angle/2)];
    
    % Create the full 8x8 gate matrix
    if qubit == 1
        % Qubit 0 (most significant)
        I = eye(2);
        gate_matrix = kron(kron(RY, I), I);
    elseif qubit == 2
        % Qubit 1 (middle)
        I = eye(2);
        gate_matrix = kron(kron(I, RY), I);
    else
        % Qubit 2 (least significant)
        I = eye(2);
        gate_matrix = kron(kron(I, I), RY);
    end
    
    new_state = gate_matrix * state;
end

function new_state = apply_controlled_ry_gate(state, control_qubit, target_qubit, angle)
    % Apply controlled RY gate
    % control_qubit: 1, 2, or 3
    % target_qubit: 1, 2, or 3
    
    % Start with identity
    gate_matrix = eye(8);
    
    % For each computational basis state, check if control is |1⟩
    for i = 1:8
        % Convert to binary representation (0-indexed)
        binary = dec2bin(i-1, 3) - '0';
        
        % Check if control qubit is 1
        if binary(control_qubit) == 1
            % Apply RY rotation to this state
            RY = [cos(angle/2), -sin(angle/2); sin(angle/2), cos(angle/2)];
            
            % Find the corresponding state with target qubit flipped
            binary_flipped = binary;
            binary_flipped(target_qubit) = 1 - binary_flipped(target_qubit);
            j = bin2dec(num2str(binary_flipped)) + 1;
            
            % Apply the rotation
            if binary(target_qubit) == 0
                gate_matrix(i, i) = cos(angle/2);
                gate_matrix(j, i) = sin(angle/2);
            else
                gate_matrix(i, i) = cos(angle/2);
                gate_matrix(j, i) = -sin(angle/2);
            end
        end
    end
    
    new_state = gate_matrix * state;
end

function new_state = apply_ccry_gate(state, control_pattern, angle)
    % Apply doubly controlled RY gate
    % control_pattern: [control1_value, control2_value] (0 or 1)
    % target is always qubit 2 (3rd qubit)
    
    gate_matrix = eye(8);
    
    for i = 1:8
        % Convert to binary representation
        binary = dec2bin(i-1, 3) - '0';
        
        % Check if both control qubits match the pattern
        if binary(1) == control_pattern(1) && binary(2) == control_pattern(2)
            % Apply RY rotation to target qubit (qubit 3)
            binary_flipped = binary;
            binary_flipped(3) = 1 - binary_flipped(3);
            j = bin2dec(num2str(binary_flipped)) + 1;
            
            if binary(3) == 0
                gate_matrix(i, i) = cos(angle/2);
                gate_matrix(j, i) = sin(angle/2);
            else
                gate_matrix(i, i) = cos(angle/2);
                gate_matrix(j, i) = -sin(angle/2);
            end
        end
    end
    
    new_state = gate_matrix * state;
end

function frequencies = simulate_measurements(state, num_shots)
    % Simulate quantum measurements
    % state: 8x1 complex vector
    % num_shots: number of measurement shots
    
    % Calculate probabilities
    probabilities = abs(state).^2;
    
    % Simulate measurements
    frequencies = zeros(1, 8);
    
    for shot = 1:num_shots
        % Random sample based on probabilities
        r = rand();
        cumulative = 0;
        
        for i = 1:8
            cumulative = cumulative + probabilities(i);
            if r <= cumulative
                frequencies(i) = frequencies(i) + 1;
                break;
            end
        end
    end
    
    % Normalize to frequencies
    frequencies = frequencies / num_shots;
end

function visualize_results(target, method1, method2)
    % Create visualization of results
    
    figure('Position', [100, 100, 800, 600]);
    
    % Create bar chart
    x = 1:8;
    width = 0.25;
    
    bar(x - width, target, width, 'FaceColor', [0.2, 0.6, 0.8], 'DisplayName', 'Target');
    hold on;
    bar(x, method1, width, 'FaceColor', [0.8, 0.4, 0.2], 'DisplayName', 'Method 1');
    bar(x + width, method2, width, 'FaceColor', [0.2, 0.8, 0.2], 'DisplayName', 'Method 2');
    
    % Customize plot
    xlabel('Quantum State');
    ylabel('Frequency');
    title('Quantum State Preparation Results');
    legend('Location', 'best');
    grid on;
    
    % Set x-axis labels
    labels = {'|000⟩', '|001⟩', '|010⟩', '|011⟩', '|100⟩', '|101⟩', '|110⟩', '|111⟩'};
    set(gca, 'XTick', 1:8, 'XTickLabel', labels);
    
    % Add error bars or annotations if needed
    ylim([0, max([target, method1, method2]) * 1.1]);
end

% Run the main function
quantum_state_preparation();